File: 2-d Array/329. Longest Increasing Path in a Matrix.js
`Given an m x n integers matrix, return the length of the longest increasing path in matrix.

From each cell, you can either move in four directions: left, right, up, or down. You may not move diagonally or move outside the boundary (i.e., wrap-around is not allowed).

 

Example 1:


Input: matrix = [[9,9,4],[6,6,8],[2,1,1]]
Output: 4
Explanation: The longest increasing path is [1, 2, 6, 9].
Example 2:


Input: matrix = [[3,4,5],[3,2,6],[2,2,1]]
Output: 4
Explanation: The longest increasing path is [3, 4, 5, 6]. Moving diagonally is not allowed.
Example 3:

Input: matrix = [[1]]
Output: 1
`;

const longestIncreasingPath = (matrix) => {
  if (matrix.length === 0) return 0;

  const rows = matrix.length;

  const cols = matrix[0].length;

  const visited = new Map();

  const dfs = (r, c, prev) => {
    const key = `${r}-${c}`;
    if (c < 0 || c === cols || r < 0 || r === rows || prev >= matrix[r][c])
      return 0;
    if (visited.has(key)) return visited.get(key);

    const result =
      1 +
      Math.max(
        dfs(r + 1, c, matrix[r][c]),
        dfs(r - 1, c, matrix[r][c]),
        dfs(r, c + 1, matrix[r][c]),
        dfs(r, c - 1, matrix[r][c])
      );
    visited.set(key, result);
    return result;
  };

  for (let row = 0; row < rows; row++) {
    for (let col = 0; col < cols; col++) {
      dfs(row, col, -1);
    }
  }
  return Math.max(...visited.values());
};

console.log(
  longestIncreasingPath([
    [1, 3, 1],
    [1, 5, 1],
    [4, 2, 1],
  ])
);

File: 2-d Array/329. Longest Increasing Path in a Matrix.py
import re


class Solution:
    def longestIncreasingPath(self, matrix) -> int:
        if not matrix or not matrix[0]:
            return 0
        rows=len(matrix)
        cols=len(matrix[0])

        dp={}
        def dfs(r,c,pre):
            if r<0 or r==rows or c<0 or c ==cols or matrix: return 0

            res=1+max(
                dfs(r-1,c,pre),
                dfs(r+1,c,pre),
                dfs(r,c-1,pre),
                dfs(r,c+1,pre)
            )
            if (r,c) not in dp:
                dp[(r,c)]=res
        
        for row in range(rows):
            for col in range(cols):
                dfs(row,col,matrix[row][col])
        print (dp)
        return max(dp.values())

        
if __name__ == '__main__':
    matrix=[[9,9,4],[6,6,8],[2,1,1]]
    print (Solution().longestIncreasingPath(matrix))